File: examples/crazyflie-edu/user_code.py
"""
User Code - Student Control Implementation

This file is where you write your control code for the Crazyflie.
The main_loop() function is called every control cycle (500 Hz by default).

The code you write here should be directly translatable to C code that
runs on the actual Crazyflie hardware (see firmware/UserCode.c).

INSTRUCTIONS:
=============
1. Read sensor data from state.gyro and state.accel
2. Check button states with state.button_blue, etc.
3. Set motor commands with state.set_motors() or state.set_all_motors()

SAFETY:
=======
- The vehicle must be ARMED for motors to spin
- The RED button acts as a dead man switch (must be held in hardware)
- Motors are automatically set to 0 when not armed

EXAMPLE (Lab 1):
================
To turn on motors when blue button is pressed:

    if state.is_armed and state.button_blue:
        state.set_all_motors(50)  # Low power for testing
    else:
        state.motors_off()

"""

from crazyflie_api import CrazyflieState

# =============================================================================
# Configuration Constants
# =============================================================================

# Team ID - Replace with your assigned team number
TEAM_ID = 0  # TODO: Set your team ID here!

# Motor command limit for safety during testing
MOTOR_TEST_LIMIT = 50  # Don't exceed this until you're confident!

# =============================================================================
# Utility Functions
# =============================================================================

# These coefficients are determined in Lab 2
# Students replace these placeholder values with their experimental fits

# PWM to Speed mapping: speed_rad_s = pwm_to_speed_a + pwm_to_speed_b * pwm
PWM_TO_SPEED_A = 426.0  # rad/s at PWM=0 (placeholder)
PWM_TO_SPEED_B = 6.8  # rad/s per PWM unit (placeholder)

# Thrust constant: force_N = THRUST_CONSTANT * speed_rad_s^2
THRUST_CONSTANT = 1.8e-8  # N/(rad/s)^2 (placeholder - identify in Lab 2!)


def pwm_from_speed(desired_speed_rad_s: float) -> float:
    """
    Convert desired motor speed (rad/s) to PWM command.

    Lab 2 Part 1: You will derive this from tachometer measurements.

    The relationship is: speed = a + b * pwm
    Inverting: pwm = (speed - a) / b

    Args:
        desired_speed_rad_s: Desired angular velocity in rad/s

    Returns:
        PWM command (0-255)
    """
    # Invert the linear relationship
    pwm = (desired_speed_rad_s - PWM_TO_SPEED_A) / PWM_TO_SPEED_B

    # Clamp to valid range
    return max(0, min(255, pwm))


def speed_from_force(desired_force_n: float) -> float:
    """
    Convert desired thrust force (N) to motor speed (rad/s).

    Lab 2 Part 2: You will derive THRUST_CONSTANT from force rig measurements.

    The relationship is: force = k * speed^2
    Inverting: speed = sqrt(force / k)

    Args:
        desired_force_n: Desired thrust per motor in Newtons

    Returns:
        Motor angular velocity in rad/s
    """
    if desired_force_n <= 0:
        return 0.0

    # Invert the quadratic relationship
    import math

    return math.sqrt(desired_force_n / THRUST_CONSTANT)


def pwm_from_force(desired_force_n: float) -> float:
    """
    Convert desired thrust force (N) directly to PWM command.

    This combines speed_from_force() and pwm_from_speed().

    Args:
        desired_force_n: Desired thrust per motor in Newtons

    Returns:
        PWM command (0-255)
    """
    speed = speed_from_force(desired_force_n)
    return pwm_from_speed(speed)


# =============================================================================
# Main Control Loop
# =============================================================================


def main_loop(state: CrazyflieState) -> None:
    """
    Main control loop - called every control cycle.

    This function is called at 500 Hz (every 2ms) during simulation.
    You have access to sensor data and can set motor commands.

    Available in 'state':
    ---------------------
    Sensors (read-only):
        state.gyro[0], state.gyro[1], state.gyro[2]  - Angular velocity (rad/s)
        state.accel[0], state.accel[1], state.accel[2] - Acceleration (g units)

    Buttons (read-only):
        state.button_blue   - Blue button pressed
        state.button_yellow - Yellow button pressed
        state.button_green  - Green button pressed
        state.button_red    - Red button (dead man switch)
        state.is_armed      - Vehicle is armed

    Timing (read-only):
        state.time          - Current time in seconds
        state.dt            - Time step in seconds

    Motor Control (write):
        state.set_motors(m1, m2, m3, m4)  - Set individual motor commands
        state.set_all_motors(cmd)         - Set all motors to same value
        state.motors_off()                - Turn all motors off

    Args:
        state: CrazyflieState object with sensor data and motor command access
    """
    # =========================================================================
    # YOUR CODE GOES HERE
    # =========================================================================

    # Lab 1 Example: Turn on motors when blue button is pressed
    # Uncomment and modify for your experiments

    # if state.is_armed and state.button_blue:
    #     # Set all motors to a low test value
    #     state.set_all_motors(MOTOR_TEST_LIMIT)
    # else:
    #     # Motors off when not testing
    #     state.motors_off()

    # =========================================================================
    # Default behavior: motors off
    # =========================================================================
    state.motors_off()


# =============================================================================
# Print Status (called when "Print Info" button is pressed)
# =============================================================================


def print_status(state: CrazyflieState) -> None:
    """
    Print diagnostic information.

    This is called when the "Print Info" button is pressed in the GUI.
    Add any debug information you want to display here.
    """
    print("=" * 50)
    print(f"Team ID: {TEAM_ID}")
    print(f"Time: {state.time:.2f} s")
    print(f"Armed: {'YES' if state.is_armed else 'NO'}")
    print()
    print("Sensor Readings:")
    print(f"  Gyro (rad/s):  x={state.gyro[0]:+.4f}, y={state.gyro[1]:+.4f}, z={state.gyro[2]:+.4f}")
    print(f"  Accel (g):     x={state.accel[0]:+.4f}, y={state.accel[1]:+.4f}, z={state.accel[2]:+.4f}")
    print()
    print("Motor Commands (PWM):")
    print(
        f"  M1={state.motor_command[0]:.0f}, M2={state.motor_command[1]:.0f}, "
        f"M3={state.motor_command[2]:.0f}, M4={state.motor_command[3]:.0f}"
    )
    print()
    print("Powertrain Constants (Lab 2):")
    print(f"  PWM_TO_SPEED_A = {PWM_TO_SPEED_A}")
    print(f"  PWM_TO_SPEED_B = {PWM_TO_SPEED_B}")
    print(f"  THRUST_CONSTANT = {THRUST_CONSTANT:.2e}")
    print("=" * 50)

File: examples/crazyflie-edu/user_code.c
/**
 * user_code.c - Student Control Implementation
 *
 * This is THE file where students write their control code.
 * The same code runs in both SITL simulation and on Crazyflie hardware.
 *
 * Part of the Crazyflie Educational Labs
 *
 * INSTRUCTIONS:
 * =============
 * 1. Set your TEAM_ID below
 * 2. Implement your control logic in user_main_loop()
 * 3. Update powertrain constants from Lab 2
 *
 * SAFETY:
 * =======
 * - Vehicle must be armed for motors to spin (Q key in sim)
 * - Blue button is dead man switch (Shift key in sim - must be held)
 * - Start with low MOTOR_TEST_LIMIT until confident
 */

#include "user_code.h"
#include <math.h>
#include <stdio.h>

// =============================================================================
// Configuration Constants
// =============================================================================

// Team ID - Replace with your assigned team number
#define TEAM_ID 0  // TODO: Set your team ID!

// Motor command limit for safety during testing
// PWM range is 0-65535; start low!
#define MOTOR_TEST_LIMIT 65000  // ~15% of max

// =============================================================================
// Powertrain Constants (Lab 2)
// =============================================================================

// PWM to Speed mapping: omega = PWM_TO_SPEED_A + PWM_TO_SPEED_B * pwm
// where pwm is in 0-255 scale (we convert from 0-65535 internally)
static const float PWM_TO_SPEED_A = 426.0f;   // rad/s at PWM=0 (placeholder)
static const float PWM_TO_SPEED_B = 6.8f;     // rad/s per PWM unit (placeholder)

// Thrust constant: force = THRUST_CONSTANT * omega^2
static const float THRUST_CONSTANT = 1.8e-8f;  // N/(rad/s)^2 (placeholder)

// Vehicle parameters
static const float VEHICLE_MASS = 0.027f;  // kg (Crazyflie 2.1)
static const float GRAVITY = 9.81f;        // m/s^2

// =============================================================================
// Utility Functions
// =============================================================================

/**
 * Convert desired motor speed (rad/s) to PWM command.
 *
 * Lab 2 Part 1: Update PWM_TO_SPEED_A and PWM_TO_SPEED_B
 * with your experimental fits.
 *
 * @param desired_speed Desired angular velocity in rad/s
 * @return PWM command (0-65535)
 */
static uint16_t pwm_from_speed(float desired_speed) {
    // Invert the linear relationship: omega = a + b * pwm
    // pwm = (omega - a) / b
    float pwm_255 = (desired_speed - PWM_TO_SPEED_A) / PWM_TO_SPEED_B;
    pwm_255 = user_clampf(pwm_255, 0.0f, 255.0f);

    // Scale from 0-255 to 0-65535
    return user_clamp_pwm(pwm_255 * 257.0f);
}

/**
 * Convert desired thrust force (N) to motor speed (rad/s).
 *
 * Lab 2 Part 2: Update THRUST_CONSTANT with your experimental value.
 *
 * @param desired_force Desired thrust per motor in Newtons
 * @return Motor angular velocity in rad/s
 */
static float speed_from_force(float desired_force) {
    // Invert the quadratic relationship: F = k * omega^2
    // omega = sqrt(F / k)
    if (desired_force <= 0.0f) {
        return 0.0f;
    }
    return sqrtf(desired_force / THRUST_CONSTANT);
}

/**
 * Convert desired thrust force (N) directly to PWM command.
 *
 * @param desired_force Desired thrust per motor in Newtons
 * @return PWM command (0-65535)
 */
static uint16_t pwm_from_force(float desired_force) {
    float speed = speed_from_force(desired_force);
    return pwm_from_speed(speed);
}

/**
 * Calculate hover thrust per motor.
 *
 * @return Thrust in Newtons for each motor to hover
 */
static float hover_thrust_per_motor(void) {
    return (VEHICLE_MASS * GRAVITY) / 4.0f;
}

// Suppress unused function warnings for utility functions
// (students will use them in later labs)
__attribute__((unused)) static uint16_t _use_pwm_from_speed = 0;
__attribute__((unused)) static float _use_speed_from_force = 0;
__attribute__((unused)) static uint16_t _use_pwm_from_force = 0;
__attribute__((unused)) static float _use_hover_thrust = 0;

// =============================================================================
// User Code Implementation
// =============================================================================

/**
 * Initialize user code - called once at startup.
 */
void user_init(void) {
    printf("[UserCode] Initialized, Team ID: %d\n", TEAM_ID);
    printf("[UserCode] Control loop: %d Hz (dt=%.3f ms)\n",
           CONTROL_LOOP_HZ, CONTROL_LOOP_DT * 1000.0f);
}

/**
 * Main control loop - called every control cycle (500 Hz).
 *
 * =========================================================================
 * YOUR CONTROL CODE GOES HERE
 * =========================================================================
 *
 * Available in 'state':
 * ---------------------
 * Sensors (read-only):
 *   state->sensors.gyro.x, .y, .z  - Angular velocity (rad/s)
 *   state->sensors.accel.x, .y, .z - Acceleration (g units)
 *
 * Buttons (read-only):
 *   state->is_armed      - Vehicle is armed
 *   state->button_blue   - Blue button (dead man switch)
 *   state->button_yellow - Yellow button
 *   state->button_green  - Green button
 *   state->button_red    - Red button
 *
 * Timing (read-only):
 *   state->time          - Current time in seconds
 *   state->dt            - Time step in seconds
 *
 * Motor Control (use helper functions):
 *   user_set_motors(state, m1, m2, m3, m4)  - Set individual motors
 *   user_set_all_motors(state, pwm)         - Set all motors same
 *   user_motors_off(state)                  - Turn all motors off
 *
 * @param state Pointer to current state
 */
void user_main_loop(user_state_t* state) {
    // Read sensor data (for reference - use as needed)
    // float gyro_x = state->sensors.gyro.x;
    // float gyro_y = state->sensors.gyro.y;
    // float gyro_z = state->sensors.gyro.z;
    // float accel_z = state->sensors.accel.z;

    // =========================================================================
    // Lab 1 Example: Turn on motors when armed AND blue button is held
    // =========================================================================
    //
    // Safety pattern:
    //   - Q key toggles "armed" state
    //   - Left Shift is the "blue button" (dead man switch - must be held)
    //   - Motors only spin when BOTH conditions are met

    if (state->is_armed && state->button_blue) {
        // Set all motors to test limit
        user_set_motors(state, MOTOR_TEST_LIMIT, MOTOR_TEST_LIMIT, MOTOR_TEST_LIMIT, MOTOR_TEST_LIMIT);
    } else {
        // Motors off - either not armed or blue button not held
        user_motors_off(state);
    }

    // =========================================================================
    // Lab 2 Example: Command specific thrust (uncomment to use)
    // =========================================================================
    /*
    if (state->is_armed && state->button_blue) {
        // Command 90% of hover thrust per motor
        float thrust = 0.9f * hover_thrust_per_motor();
        uint16_t pwm = pwm_from_force(thrust);
        user_set_all_motors(state, pwm);
    } else {
        user_motors_off(state);
    }
    */
}

/**
 * Print status information - called when "Print Info" is requested.
 *
 * @param state Pointer to current state
 */
void user_print_status(const user_state_t* state) {
    printf("================================================\n");
    printf("User Code Status - Team %d\n", TEAM_ID);
    printf("================================================\n");
    printf("Time: %.2f s\n", state->time);
    printf("Armed: %s\n", state->is_armed ? "YES" : "NO");
    printf("\n");
    printf("Sensors:\n");
    printf("  Gyro (rad/s):  x=%+.4f, y=%+.4f, z=%+.4f\n",
           state->sensors.gyro.x, state->sensors.gyro.y, state->sensors.gyro.z);
    printf("  Accel (g):     x=%+.4f, y=%+.4f, z=%+.4f\n",
           state->sensors.accel.x, state->sensors.accel.y, state->sensors.accel.z);
    printf("\n");
    printf("Buttons: Blue=%d Yellow=%d Green=%d Red=%d\n",
           state->button_blue, state->button_yellow,
           state->button_green, state->button_red);
    printf("\n");
    printf("Motors (PWM):\n");
    printf("  M1=%5u  M2=%5u  M3=%5u  M4=%5u\n",
           state->motor_pwm[0], state->motor_pwm[1],
           state->motor_pwm[2], state->motor_pwm[3]);
    printf("\n");
    printf("Powertrain Constants (Lab 2):\n");
    printf("  PWM_TO_SPEED_A = %.2f rad/s\n", PWM_TO_SPEED_A);
    printf("  PWM_TO_SPEED_B = %.4f rad/s per PWM\n", PWM_TO_SPEED_B);
    printf("  THRUST_CONSTANT = %.2e N/(rad/s)^2\n", THRUST_CONSTANT);
    printf("\n");
    printf("Calculated:\n");
    printf("  Hover thrust/motor = %.4f N\n", hover_thrust_per_motor());
    printf("  Hover speed = %.1f rad/s\n", speed_from_force(hover_thrust_per_motor()));
    printf("  Hover PWM = %u\n", pwm_from_force(hover_thrust_per_motor()));
    printf("================================================\n");
}

